vowpalwabbit/oaa.cc

```diff
@@ -264,51 +264,70 @@ VW::LEARNER::base_learner* oaa_setup(VW::setup_base_i& stack_builder)
   {
+    pred_type = prediction_type_t::scalars;
     if (probabilities)
     {
       auto loss_function_type = all.loss->getType();
       if (loss_function_type != "logistic")
         *(all.trace_message) << "WARNING: --probabilities should be used only with --loss_function=logistic"
                              << std::endl;
       // the three boolean template parameters are: is_learn, print_all and scores
-      l = &LEARNER::init_multiclass_learner(data, base, learn<false, true, true>, predict<false, true, true>,
-          all.example_parser, data->k, stack_builder.get_setupfn_name(oaa_setup) + "-prob", prediction_type_t::scalars);
-      all.example_parser->lbl_parser.label_type = label_type_t::multiclass;
+      learn_ptr = learn<false, true, true>;
+      pred_ptr = predict<false, true, true>;
+      name_addition = "-prob";
+      finish_ptr = finish_example_scores<true>;
       all.sd->report_multiclass_log_loss = true;
-      l->set_finish_example(finish_example_scores<true>);
     }
     else
     {
-      l = &VW::LEARNER::init_multiclass_learner(data, base, learn<false, true, false>, predict<false, true, false>,
-          all.example_parser, data->k, stack_builder.get_setupfn_name(oaa_setup) + "-scores",
-          prediction_type_t::scalars);
-      all.example_parser->lbl_parser.label_type = label_type_t::multiclass;
-      l->set_finish_example(finish_example_scores<false>);
+      learn_ptr = learn<false, true, false>;
+      pred_ptr = predict<false, true, false>;
+      name_addition = "-scores";
+      finish_ptr = finish_example_scores<false>;
     }
   }
-  else if (all.raw_prediction != nullptr)
-  {
-    l = &VW::LEARNER::init_multiclass_learner(data, base, learn<true, false, false>, predict<true, false, false>,
-        all.example_parser, data->k, stack_builder.get_setupfn_name(oaa_setup) + "-raw", prediction_type_t::multiclass);
-    all.example_parser->lbl_parser.label_type = label_type_t::multiclass;
-  }
   else
   {
-    l = &VW::LEARNER::init_multiclass_learner(data, base, learn<false, false, false>, predict<false, false, false>,
-        all.example_parser, data->k, stack_builder.get_setupfn_name(oaa_setup), prediction_type_t::multiclass);
-    all.example_parser->lbl_parser.label_type = label_type_t::multiclass;
+    pred_type = prediction_type_t::multiclass;
+    finish_ptr = MULTICLASS::finish_example<oaa>;
+    if (all.raw_prediction != nullptr)
+    {
+      learn_ptr = learn<true, false, false>;
+      pred_ptr = predict<true, false, false>;
+      name_addition = "-raw";
+    }
+    else
+    {
+      learn_ptr = learn<false, false, false>;
+      pred_ptr = predict<false, false, false>;
+      name_addition = "";
+    }
   }
 
+  all.example_parser->lbl_parser = MULTICLASS::mc_label;
+
   if (data_ptr->num_subsample > 0)
   {
-    l->set_learn(learn_randomized);
-    l->set_finish_example(MULTICLASS::finish_example_without_loss<oaa>);
-    l->learn_returns_prediction = false;
```

Comment:
`learn_returns_prediction` which makes this line a noop. Was this intentional? @rajan-chari

Reply:
The semantics and name of that variable was reversed so this is likely vestigial and can be removed.
There are reductions with conditional behavior for learn_returns_predicttion where being explicit reduces cognitive burden.